### PR TITLE
Translation of LLVM intrinsics

### PR DESCRIPTION
**Pull request by AlexeySotkin**
An LLVM intrinsic should not be transalated to SPIRV as a function call.
Either it's semantics is represented via SPIRV instruction(s)
or an error is reported.

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -1201,6 +1220,63 @@ LLVMToSPIRV::transSpcvCast(CallInst* CI, SPIRVBasicBlock *BB) {
...
+      MemoryAccess[0] |= MemoryAccessAlignedMask;
+      MemoryAccess.push_back(AlignVal);
+    }
+    if (MI->isVolatile())
+      MemoryAccess[0] |= MemoryAccessVolatileMask;
+    return MemoryAccess;
+  };
+
+  switch (II->getIntrinsicID()) {
+  case Intrinsic::fmuladd : {
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVWriter.cpp:1236]: Probably we should translate llvm.fma.\* the same way and emit a warning.

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -1201,6 +1220,63 @@ LLVMToSPIRV::transSpcvCast(CallInst* CI, SPIRVBasicBlock *BB) {
...
+                                        BB);
+    return BM->addBinaryInst(OpFAdd, Ty, Mul,
+                             transValue(II->getArgOperand(2), BB), BB);
+  }
+  case Intrinsic::memset : {
+    // If memset is used for zero-initialization, find a variable which is being
+    // initialized and store null constant of the same type to this variable
+    MemSetInst *MSI = cast<MemSetInst>(II);
+    ConstantInt *Val = dyn_cast<ConstantInt>(MSI->getValue());
+    if (Val && Val->isZero()) {
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVWriter.cpp]: This only works if the size and alignment of the pointee type matches the length and alignment argument of memset instruction. This is not always true, so need to add these two conditions.

File: lib/SPIRV/SPIRVWriter.cpp (hunk 1 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVWriter.cpp:1236]: We don't have immediate need for this right now, do we? Besides there is no facilities for emitting warnings. This commit is not intended to cover all llvm intrinsics, I just wanted to fix #174 and llvm.memset which revealed during testing. Probably we should submit another issue to handle llvm.fma.*

File: lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVWriter.cpp]: Fixed

File: lib/SPIRV/SPIRVWriter.cpp (hunk 1 above)
Comment (yxsamliu) [reply lib/SPIRV/SPIRVWriter.cpp:1236]: OK. Let's leave it until there is need to translate this.

**Pull request merged**